cloud_storage: use `topic_namespace_override` in `remote_path_provider`

WillemKauf opened the pull request:
This PR allows for the `remote_path_provider` to use a `topic_namespace_override` from `topic_properties` when constructing remote paths to segments, manifests, and other files.

The motivating use-case for this work is cross-cluster migrations, in which a destination cluster can refer to a migrated topic (`src`) by an alias (`dest`) for its local storage, while continuing to use the existing topic name (`src`) for the cloud storage read and write paths.

It is not currently possible to create a topic with this override through `rpk`/ the `kafka` subsystem.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

* none

WillemKauf commented:
Force push to correct `compat` related tests, as well as `offline_log_viewer` functionality with the added property in `cluster::topic_properties`.

WillemKauf commented:
CI failure is https://github.com/redpanda-data/redpanda/issues/18563

vbotbuildovich commented:
<!--NO CACHE-->
skipped ducktape retry in https://buildkite.com/redpanda/redpanda/builds/51117#01907ffa-44fe-4f4d-aaba-c06f98f39dc2:
pandatriage cache was not found
<!--3c102db5-dbee-4e11-884a-53e501b9988c-->

WillemKauf commented:
Force push to remove unnecessary additions to dead `adl<topic_properties>` code.

WillemKauf commented:
Force push in order to rebase to upstream `dev`.

* Changed behaviour in `ntp_archiver`, where `archiver_service::_ntp` now reflects the "local" `ntp` once more.
* Added `ntp_archiver::get_archival_ntp()`, which returns `manifest().get_ntp()` (which is assigned to be the `_raft->archival_ntp()` upon construction)

Still a WIP in terms of figuring out the best way to override the path for tiered storage without muddying the definition of an `ntp`, per discussions with Nicolae.

WillemKauf commented:
Force push to rebase to upstream `dev`.

WillemKauf commented:
Force push after reworking PR. Good for review.

WillemKauf commented:
Failure is Bazel build.

andrwng commented on src/v/cloud_storage/tests/remote_path_provider_test.cc:
```diff
@@ -85,84 +129,123 @@ TEST(TopicFromPathTest, TestTopicFromPrefixedTopicManifestPath) {
-      "e0000000/meta/kafka/tp/5_21/manifest.bin");
+      fmt::format("{}/5_21/manifest.bin", expected_path).c_str());
 
     spillover_manifest spill_m(test_ntp, test_rev);
     spill_m.add(test_smeta);
     EXPECT_STREQ(
       path_provider.partition_manifest_path(spill_m).c_str(),
-      "e0000000/meta/kafka/tp/5_21/manifest.bin.10.11.0.1.999.1000");
+      fmt::format("{}/5_21/manifest.bin.10.11.0.1.999.1000", expected_path)
+        .c_str());
     EXPECT_STREQ(
       spill_m.get_manifest_path(path_provider)().native().c_str(),
-      "e0000000/meta/kafka/tp/5_21/manifest.bin.10.11.0.1.999.1000");
+      fmt::format("{}/5_21/manifest.bin.10.11.0.1.999.1000", expected_path)
+        .c_str());
     partition_manifest* disguised_manifest = &spill_m;
     EXPECT_STREQ(
       disguised_manifest->get_manifest_path(path_provider)().native().c_str(),
-      "e0000000/meta/kafka/tp/5_21/manifest.bin.10.11.0.1.999.1000");
+      fmt::format("{}/5_21/manifest.bin.10.11.0.1.999.1000", expected_path)
+        .c_str());
 }
 
-TEST(RemotePathProviderTest, TestLabeledPartitionManifestPaths) {
-    remote_path_provider path_provider(test_label, test_tp_ns_override);
+TEST_P(
+  OverrideParamRemotePathProviderWithLabelTest,
+  TestLabeledPartitionManifestPaths) {
+    const auto& expected_ns = has_override ? test_tp_ns_override.ns
+                                           : test_tp_ns.ns;
+    const auto& expected_tp = has_override ? test_tp_ns_override.tp
+                                           : test_tp_ns.tp;
+    const auto expected_path = fmt::format(
+      "meta/{}/{}", expected_ns, expected_tp);
     EXPECT_STREQ(
       path_provider.partition_manifest_path(test_ntp, test_rev).c_str(),
-      "deadbeef-0000-0000-0000-000000000000/meta/kafka/tp/5_21/manifest.bin");
+      fmt::format(
+        "deadbeef-0000-0000-0000-000000000000/{}/5_21/manifest.bin",
+        expected_path)
```
nit: as far as reading/understanding the path provider, i liked the fact that previously we could see exactly some unobstructed examples of the paths that were being generated. I know it will be more code, but would you mind writing the expected paths as plain strings? Especially for something that isn't complicated logic, IMO that kind of readability trumps reuse

WillemKauf replied:
Rewrote this test with more explicit path strings.

andrwng commented on src/v/cloud_storage/tests/topic_manifest_test.cc:
```diff
@@ -459,6 +459,7 @@ SEASTAR_THREAD_TEST_CASE(test_topic_manifest_serde_feature_table) {
       true,
       "rrr",
       std::nullopt,
+      std::nullopt,
```
nit: commit message needs an update now that there is no archival_ntp

WillemKauf replied:
Updated commit message.

andrwng commented on src/v/cloud_storage/remote_path_provider.cc:
```diff
@@ -32,30 +32,37 @@ remote_path_provider remote_path_provider::copy() const {
 
 ss::sstring remote_path_provider::topic_manifest_prefix(
   const model::topic_namespace& topic) const {
+    const auto& tp_ns = _topic_namespace_override.value_or(topic);
```
nit: please squash this with the commit that adds tests for the remote_path_provider

WillemKauf replied:
Squashed commits.

andrwng commented on src/v/cloud_storage/remote_path_provider.cc:
```diff
@@ -32,30 +32,37 @@ remote_path_provider remote_path_provider::copy() const {
 
 ss::sstring remote_path_provider::partition_manifest_prefix(
-  const model::ntp& ntp, model::initial_revision_id rev) const {
+  model::ntp ntp, model::initial_revision_id rev) const {
```
I doubt it's worth optimizing, but it's a shame this can't avoid copies in the common path. Ultimately it probably doesn't matter anyway since this is generating a brand new string

WillemKauf replied:
Yeah, I was looking for a way to avoid copies, but couldn't really find the best way to do so here, with the exception of adding some sort of `ntp_view` type, or changing around the call sites to use `topic_namespace_view/partition` as arguments while adding some new `path()` functions that accept those types as well.

Future work, perhaps.

andrwng reviewed (commented):
Looking good! It'd be great if we could also add a test (maybe in a new test file) that wipes a cluster and restores the topic under the alias, and then uses the Kafka to read from it.

TestRecoveryControllerState in cluster_recovery_backend_test.cc has some examples of wiping a cluster and rebooting.

WillemKauf commented:
> It'd be great if we could also add a test (maybe in a new test file) that wipes a cluster and restores the topic under the alias, and then uses the Kafka to read from it.

Great callout, added `topic_namespace_override_recovery_test.cc` that covers this case.

WillemKauf commented:
Force push to rebase to upstream `dev` to fix failing build (additions in `src/v/archival/tests/archiver_operations_impl_gtest.cc`)

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/51667#0190c1c7-ac73-47a9-b3db-1d34fe9acd8e
<!--b0865866-c174-4e1f-aa78-63c58c14055c-->

WillemKauf commented on src/v/cluster/tests/topic_namespace_override_recovery_test.cc:
Done

WillemKauf commented on src/v/cluster/topic_configuration.h:
```diff
@@ -48,6 +48,16 @@ struct topic_configuration
+            return properties.topic_namespace_override.value();
+        }
+        return tp_ns;
+    }
```
I think I may end up using these in the future elsewhere- if you feel strongly about removing them, I can do so.

For now, I renamed to `remote_topic_namespace_override` and `remote_tp_ns`.

WillemKauf commented on src/v/cluster/tests/CMakeLists.txt:
```diff
@@ -206,6 +206,22 @@ rp_test(
+    v::storage_test_utils
+  ARGS "-- -c 1"
+  LABELS cluster
+)
```
Agreed, I think I was just working in the `cluster` directory at the time, and could have gone either way with the location.

Moved the test to `gtest_cloud_storage_rpfixture`.

andrwng commented on src/v/cloud_storage/tests/topic_namespace_override_recovery_test.cc:
Does this actually fail the test if it fires? Aren't we ignoring the results of the uplaod below? I think it's more foolproof if we just loop through `get_requests()` and assert on contains("tapioca")

WillemKauf replied:
There will be parsing errors from the `xml_parser` in `s3_client` if this fires due to the invalid response body, and the test will fail. I suppose this is isn't as direct a failure as asserting over requests in `get_requests()`, I can make the change if you have a strong preference for it.

andrwng replied:
I think it's a worthwhile (and hopefully simple?) change. The current code assumes that the only things hitting cloud are triggered by this test and that such exceptions aren't handled/ignored by the upload path. In practice I think it doesn't really make a difference since I dont think we're running housekeeping or anything, but just pointing out why this is IMO not as airtight a test

andrwng commented on src/v/cloud_storage/tests/topic_namespace_override_recovery_test.cc:
Probably worth fixing the interface since it's easy to make this mistake: num_segment produces until the total number of segments is >=2, rather than adding 2 segments. This value should be >=10 (worth double checking in your local runs that that's the case)

WillemKauf replied:
Yeah, this was intended on my end- but if `10` is a safer value than `2`, I'm happy to update it.
